gson/src/main/java/com/google/gson/ToNumberPolicy.java

```diff
@@ -68,22 +68,30 @@ public Number readNumber(JsonReader in) throws IOException {
     @Override
     public Number readNumber(JsonReader in) throws IOException, JsonParseException {
       String value = in.nextString();
-      try {
-        return Long.parseLong(value);
-      } catch (NumberFormatException longE) {
+      if (value.indexOf('.') >= 0) {
+        return parseAsDouble(value, in);
+      } else {
         try {
-          Double d = Double.valueOf(value);
-          if ((d.isInfinite() || d.isNaN()) && !in.isLenient()) {
-            throw new MalformedJsonException(
-                "JSON forbids NaN and infinities: " + d + "; at path " + in.getPreviousPath());
-          }
-          return d;
-        } catch (NumberFormatException doubleE) {
-          throw new JsonParseException(
-              "Cannot parse " + value + "; at path " + in.getPreviousPath(), doubleE);
+          return Long.parseLong(value);
+        } catch (NumberFormatException longE) {
```

Comment:
Since you've moved the double parsing into a separate method (which I think is a good change), you no longer need to distinguish between `longE` and `doubleE`. I would suggest just calling them both `e`.

Reply:
@eamonnmcmanus good call. Done